### ISSUE TITLE
GPT Neo/J Padding Side Results in Different Generation Outputs

### ISSUE THREAD

**Issue by anas-awadalla** [huggingface/transformers#16595]
## Environment info

- `transformers` version: 4.17.0
- Platform: Linux-5.4.144+-x86_64-with-Ubuntu-18.04-bionic
- Python version: 3.7.13
- PyTorch version (GPU?): 1.10.0+cu111 (False)
- Tensorflow version (GPU?): 2.8.0 (False)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: No
- Using distributed or parallel set-up in script?: None

### Who can help
@patil-suraj @patrickvonplaten 

## Information
I find that for GPT Neo and GPT J the padding side has a great effect on the quality of the generation. More specifically it seems that while padding on the left results in decent generation output, padding on the right causes the model to generate empty spaces or repeating characters (ex. "AAAAAAA"). I am wondering why padding has this effect on generation.

Model I am using (Bert, XLNet ...): GPT Neo 125M (although I also observed the same issue for larger variants of Neo and GPT-J)

The problem arises when using:
* [ ] the official example scripts: (give details below)
* [x] my own modified scripts: (give details below)

The tasks I am working on is:
* [ ] an official GLUE/SQUaD task: (give the name)
* [x] my own task or dataset: (give details below)

## To reproduce

Run this [Colab](https://colab.research.google.com/drive/1LdJ4kt_egJprxmpWZJSwkbvx71kN5RFv?usp=sharing) 

or

1. Load any of the GPT Neo or GPT J models
2. Compare the results of padding inputs on the left vs right on generation outcome

## Expected behavior
I expect that it shouldn't matter whether we pad the input from the left or right but it seems like that has a effect on generation results.

**Comment by patil-suraj** [comment]
Hi @anas-awadalla !

This is because these are auto-regressive models, which generate next tokens based on previous tokens. So if the inputs are padded to the right the model will receive padded tokens as context, which is not useful during generation. 

Because of this for generation with any auto-regressive models, the input is padded to the right.

**Comment by anas-awadalla** [comment]
Thanks for your reply! Did you mean the input should be padded to the left?

**Comment by anas-awadalla** [comment]
> Thanks for your reply! Did you mean the input should be padded to the left? In both cases the input is passed in as part of the context so can you clarify why one padding strategy is better? Also in the generate function we pass input ids and an attention mask, shouldn’t the attention mask indicate what parts of the input is padding tokens and the model should not consider those as part of the context?

Bumping this :)

**Comment by ZonglinY** [comment]
I meet the same problem... Even padding to the left still results in worse performance than not padding.

**Comment by anas-awadalla** [comment]
@patil-suraj Could you clarify your response above? I closed the issue previously because padding on the left works well for me but would still like to understand why that's the case.
